src/main/java/org/myrobotlab/service/Hd44780I2c.java

```diff
@@ -0,0 +1,242 @@
+package org.myrobotlab.service;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.myrobotlab.framework.Platform;
+import org.myrobotlab.framework.Service;
+import org.myrobotlab.framework.ServiceType;
+import org.myrobotlab.logging.Level;
+import org.myrobotlab.logging.LoggerFactory;
+import org.myrobotlab.logging.LoggingFactory;
+import org.slf4j.Logger;
+
+/**
+ * 
+ * HD44780 - Driver to play with lcd display panel
+ * Tested with HD44780 1602 panel, attached to pcf8574
+ * Driver pasted from Poduzov : https://github.com/Poduzov/PI4J-I2C-LCD
+ * 
+ * @author Moz4r
+ * 
+ */
+public class Hd44780I2c extends Service {
+
+  public Hd44780I2c(String reservedKey) {
+    super(reservedKey);
+    setReady(false);
+  }
+
+  public final static Logger log = LoggerFactory.getLogger(Hd44780I2c.class);
+  private static final long serialVersionUID = 1L;
+
+  private Pcf8574 pcf8574;

```

Comment:
This device should be able to be attached to any I2CController service.  Can we re-write this to use the Interface, rather than hard coding that it needs to be attached to a Pcf8574 ?

Reply:
No, I tried, pcf8574 don't use I2CController interface , it is not a I2CController service but a I/O bus extender for i2c bus. 
I'm not sure if we can get it worky with other stuff inside mrl ... maybe later

Reply:
The PCF8574 is a i2c bus extender. It will give 8 extra I/O pins. The pcf8574 is built to be able to connect to any I2cController. So it can be used on the Arduino, RasPi or ESP8266_01 already. So that goal is already acheived.

Reply:
Who's going to write the interface ? :)

Reply:
I was thinking that the display LcdDisplay is a standard i2c device.  it, thus should be an I2CControl and should be attachable to an I2CController.   We don't have to do it now, but I think this is a direction we should go in, for future i2c devices.

Reply:
Hi. 
There is two different devices involved here:

One is the LCD display driver board HD44780. It has no i2c interface. It can potentially also be connected directly to the pins of an Arduino or RasPi.
The service moz4r has built is a represenation of that board. The reason he is using the service Pcf8574 directly, and not using an interface, is because I never put the methods that you can use to communicate with it in a generic port extender interface. I think I should do that, because then moz4r could use that interface.

The Pcf8574 is an i2c device, and implements the I2CControl interface and can talk to any i2c controller using the I2CController interface. So I think we are good there.
It also implements PinArrayControl, and thats how I was expecting it to be used. That works fine and for example the Pcf8574Gui use the same Oscope component that is used in the ArduinoGui. But I will add a generic PortExtenderControl interface that moz4 can use, because that is the missing piece of that service. 

If LcdDisplay (HD44780) indeed was an i2c device, I would agree 100% with what Kwatters write. But in this case the pcf8574 is the i2c device and LcdDisplay (HD44780) is not.
The alternative would be to look at the combination of HD44780 and pcf8574 as a single unit. In that case I would also agree with Kwatters. Perhaps that is a better way to go. Then no new interface would be needed, but all i2c configuration ( I2CControl ) would need to be implemented in that service. 
And I think some generic LcdDisplayControl interface should be defined for the LCD methods.

Reply:
Sounds good - would you like to write another ticket(s) for the missing work ?
All you have more experience than myself for these components ... I guess you still have a couple of decisions to go through but possible tickets (as I understand it would be)

- create generalized LcdDisplayControl interface
- create PortExtenderControl interface - integrate it with this service (or not?) and implement I2CControl in this service